### PR TITLE
[#234] Throw more and better failure messages on storage creation

### LINKED ISSUE #234: Throw more and better failure messages on storage creation
# Clarification and motivation

In most cases we handle storage generation checks (e.g. on bounds) silently, by making sure that the necessary properties are respected.

This is not very good as it's not clear from the command execution that something required a change, we should use `validate_*` functions instead and throw an error message instead when a change is necessary.

Moreover, we know from the start that some of the `Makefile` target parameters should be provided (e.g. the `admin` address), however we provide default values for them.

This also is likely a bad idea because it may not make very clear that something is necessary. 

# Acceptance criteria

1. which of the `Makefile` storage targets argument should always be provided by the user is established and default values are no longer provided for these
2. storage validation checks are handled with failure messages and no longer with silent changes
3. the `building` documentation is updated accordingly.

Note: because this is storage generation and they will not end up in contract `code`, messages can be as long and expressive as we want.

### PR DESCRIPTION
**Pull request by rinn7e**
## Description

Problem: Currently, we provide `Makefile` arugments by default and we
handle some storage check silently. This is not really a good idea.

Solution: Remove default values in `Makefile`, and throw error instead.
Add more validate function to storage generation.

## Related issue(s)

Resolves #234 

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by rinn7e** [comment]
@pasqu4le Continue from the discussion from here: https://github.com/tqtezos/baseDAO/pull/248#issuecomment-845915783

I've made the changes according to your review, and it works fine now. I'm not sure what are the other validate functions to put in, currently I only add `validate_quorum_threshold` check for now. 

Hope to get your review on this.

File: Makefile [hunk 1]
```diff
@@ -24,6 +24,8 @@ BUILD_PARAMETER = $(LIGO) compile-parameter --syntax cameligo
 escape_quote = $(subst ','\'',$(1))
 # Utility function to escape double quotes
 escape_double_quote = $(subst $\",$\\",$(1))
+# Utility function to require a variable to be defined
+require_defined = $(if $(value $1),$(value $1),$(error "$1 needs to be defined"))
```
Comment (sras) [inline Makefile:28]: Does this work if `$1` is zero?

Comment (rinn7e) [reply Makefile:28]: It seems to work fine if `$1` is `0`. It only throws error if the value `$1` is an empty string.

**Comment by sras** [comment]
@rinn7e Not sure, but it appears that we still have default values for trivialDAO in the Makefile. Shouldn't we remove it as well?

https://github.com/tqtezos/baseDAO/blob/a04d1a3a735af25017d00e01b6e2c99c30fdb47d/Makefile#L59

**Comment by rinn7e** [comment]
> @rinn7e Not sure, but it appears that we still have default values for trivialDAO in the Makefile. Shouldn't we remove it as well?
> 
> https://github.com/tqtezos/baseDAO/blob/a04d1a3a735af25017d00e01b6e2c99c30fdb47d/Makefile#L59

My bad, forgot to remove this. I've pushed an update now.

**Review by sras — APPROVED** [review]
LGTM now

File: Makefile [hunk 2]
```diff
@@ -57,188 +56,196 @@ ifeq ($(OPTIMIZE), true)
...
-$(OUT)/trivialDAO_storage.tz : ledger = ([] : ledger_list)
-$(OUT)/trivialDAO_storage.tz : quorum_threshold = 10n
-$(OUT)/trivialDAO_storage.tz : min_quorum = 1n
-$(OUT)/trivialDAO_storage.tz : max_quorum = 99n
-$(OUT)/trivialDAO_storage.tz : period = 950400n # 11 days
-$(OUT)/trivialDAO_storage.tz : quorum_change = 5n
-$(OUT)/trivialDAO_storage.tz : max_quorum_change = 19n
-$(OUT)/trivialDAO_storage.tz : governance_total_supply = 100n
-$(OUT)/trivialDAO_storage.tz : proposal_flush_time = 1900801n # 22 days 1 seconds
-$(OUT)/trivialDAO_storage.tz : proposal_expired_time = 2851200n # 33 days
```
Comment (pasqu4le) [inline Makefile:76]: Oh, you ended up making them all required, I was thinking of removing the default value only for some of them.

See the issue description again, I think it's pretty clear that what fields to require should be decided on first, especially from the first acceptance criteria:
> which of the Makefile storage targets argument should always be provided by the user is established and default values are no longer provided for these

I don't think they all fit the condition that they "should be provided" by the user.

I mean, my point is that with the default e.g. `admin_address` or `guardian_address` a real contract won't actually make sense, so the user should provide them in any case.
However, values like the one of `period` or `quorum_change` have decent default values which can be used for a real deployment, so we can keep giving defaults that can be overridden.

File: src/defaults.mligo [hunk 3]
```diff
@@ -4,15 +4,21 @@
 #include "types.mligo"
 #include "proposal.mligo"
```
Comment (pasqu4le) [inline src/defaults.mligo:6]: One more thing to check is that the `quorum_threshold` is between the `min_quorum_threshold` and `max_quorum_threshold` values.

Comment (rinn7e) [reply src/defaults.mligo:6]: I've pushed an update to this now.

File: Makefile (hunk 2 above)
Comment (pasqu4le) [reply Makefile]: No problem, perhaps I should have made it more clear.

Yes, I think those should be all the required values :+1:

Comment (rinn7e) [reply Makefile]: If that's the case, I've pushed the update. You can take a look at it.

File: Makefile [hunk 4]
```diff
@@ -203,42 +190,62 @@ $(OUT)/treasuryDAO_storage.tz: src/**
...
+          ; max_xtz_amount = ($(call require_defined,max_xtz_amount) : tez) \
           })"
 
 	# ============== Compilation successful ============== #
 	# See "$(OUT)/treasuryDAO_storage.tz" for compilation result #
 	#
 
-test: all
+# For development and testing purpose
+storage:
```
Comment (pasqu4le) [inline Makefile:229]: I suggest renaming this target `test-storage` instead

File: Makefile [hunk 5]
```diff
@@ -203,42 +190,62 @@ $(OUT)/treasuryDAO_storage.tz: src/**
...
-          ; slash_division_value = $(slash_division_value) \
-          ; min_xtz_amount = $(min_xtz_amount) \
-          ; max_xtz_amount = $(max_xtz_amount) \
+          ; frozen_scale_value = ($(call require_defined,frozen_scale_value) : nat) \
+          ; frozen_extra_value = ($(call require_defined,frozen_extra_value) : nat) \
+          ; max_proposal_size = ($(call require_defined,max_proposal_size) : nat) \
+          ; slash_scale_value = ($(call require_defined,slash_scale_value) : nat) \
+          ; slash_division_value = ($(call require_defined,slash_division_value) : nat) \
+          ; min_xtz_amount = ($(call require_defined,min_xtz_amount) : tez) \
+          ; max_xtz_amount = ($(call require_defined,max_xtz_amount) : tez) \
```
Comment (pasqu4le) [inline Makefile:221]: These changes to use `require_defined` are now unnecessary.

Since there is a default value there will never be a case in which there are undefined.

I mean, it could only happen if one purposefully calls e.g. `make out/treasuryDAO_storage.tz max_xtz_amount=""`, but LIGO will throw a better error in that case anyway.

Note: this comment applies to the other non-necessary changes like this, including the other storage targets.

File: Makefile [hunk 6]
```diff
@@ -184,9 +171,9 @@ $(OUT)/treasuryDAO_storage.tz : slash_division_value = 0n
 $(OUT)/treasuryDAO_storage.tz : min_xtz_amount = 0mutez
 $(OUT)/treasuryDAO_storage.tz : max_xtz_amount = 100mutez
 $(OUT)/treasuryDAO_storage.tz : now_val = `date +"%s"`
-$(OUT)/treasuryDAO_storage.tz : metadata_map = (Big_map.empty : metadata_map)
+$(OUT)/treasuryDAO_storage.tz : metadata_map = Big_map.empty
 $(OUT)/treasuryDAO_storage.tz : fixed_proposal_fee_in_token = 0n
-$(OUT)/treasuryDAO_storage.tz : ledger = ([] : ledger_list)
+$(OUT)/treasuryDAO_storage.tz : ledger = []
```
Comment (pasqu4le) [inline Makefile:176]: Even if you revert the changes below (as I suggested in the other comment) to remove the use of `require_defined` that are not necessary, I think we should keep this (and move only the type annotation in the call below).

It seems to me it makes things cleaner overall.

File: src/defaults.mligo [hunk 7]
```diff
@@ -4,15 +4,22 @@
...
   else if data.proposal_flush_time <= (data.period.length * 2n) then
-    failwith("proposal_flush_time needs to be more than twice the period length")
+    failwith("'proposal_flush_time' needs to be more than twice the 'period' length.")
   else unit
 
+let validate_quorum_threshold_bound (data : initial_config_data) : unit =
+  if data.max_quorum >= data.quorum_threshold && data.quorum_threshold >= data.min_quorum then
+    unit
+  else
+    failwith("'quorum_threshold' value need be between 'max_quorum' and 'min_quorum'.")
```
Comment (pasqu4le) [inline src/defaults.mligo:18]: Let's be as detailed as possible here, I think we should split the check in two and give two different messages for "greater or equal then `min_quorum`" and "less than or equal to `max_quorum`".

File: Makefile (hunk 5 above)
Comment (rinn7e) [reply Makefile]: I kept it in case when we need it in the future. But I'm good with removing it for now. I pushed the update and handled other changes as well.

File: Makefile [hunk 8]
```diff
@@ -203,42 +190,62 @@ $(OUT)/treasuryDAO_storage.tz: src/**
...
+		governance_token_address=KT1RdwP8XJPjFyGoUsXFQnQo1yNm6gUqVdp5 \
+		governance_token_id=0n
+
+	make $(OUT)/registryDAO_storage.tz \
+		admin_address=tz1QozfhaUW4wLnohDo6yiBUmh7cPCSXE9Af \
+		guardian_address=KT1QbdJ7M7uAQZwLpvzerUyk7LYkJWDL7eDh \
+		governance_token_address=KT1RdwP8XJPjFyGoUsXFQnQo1yNm6gUqVdp5 \
+		governance_token_id=0n
+
+test: storage all
```
Comment (pasqu4le) [inline Makefile:248]:
```suggestion
test: test-storage all
```

**Review by pasqu4le — APPROVED** [review]
Only one last minor thing, but LGTM now
